integration/test/Test/Login.hs

```diff
@@ -0,0 +1,115 @@
+  (owner, team, []) <- createTeam OwnDomain 0
+  email <- owner %. "email"
+  setTeamFeatureLockStatus owner team "sndFactorPasswordChallenge" "unlocked"
+  setTeamFeatureStatus owner team "sndFactorPasswordChallenge" "enabled"
+  bindResponse (login owner email defPassword) $ \resp -> do
+    resp.status `shouldMatchInt` 403
+    resp.json %. "label" `shouldMatch` "code-authentication-required"
+
+-- @END
+
+-- @SF.Channel @TSFI.RESTfulAPI @S2
+--
+-- Test that login fails with expired second factor email verification code
+testLoginVerify6DigitExpiredCodeFails :: HasCallStack => App ()
+testLoginVerify6DigitExpiredCodeFails = do
+  withModifiedBackend
+    (def {brigCfg = setField "optSettings.setVerificationTimeout" (Aeson.Number 2)})
+    $ \domain -> do
+      (owner, team, []) <- createTeam domain 0
+      email <- owner %. "email"
+      setTeamFeatureLockStatus owner team "sndFactorPasswordChallenge" "unlocked"
+      setTeamFeatureStatus owner team "sndFactorPasswordChallenge" "enabled"
+      generateVerificationCode owner email
+      code <- getVerificationCode owner "login" >>= getJSON 200 >>= asString
+      liftIO $ threadDelay 2_000_100
+      bindResponse (loginWith2ndFactor owner email defPassword code) \resp -> do
+        resp.status `shouldMatchInt` 403
+        resp.json %. "label" `shouldMatch` "code-authentication-failed"
+
+-- @END
+
+testLoginVerify6DigitResendCodeSuccessAndRateLimiting :: HasCallStack => App ()
+testLoginVerify6DigitResendCodeSuccessAndRateLimiting = do
+  (owner, team, []) <- createTeam OwnDomain 0
+  email <- owner %. "email"
+  setTeamFeatureLockStatus owner team "sndFactorPasswordChallenge" "unlocked"
+  setTeamFeatureStatus owner team "sndFactorPasswordChallenge" "enabled"
+  generateVerificationCode owner email
+  fstCode <- getVerificationCode owner "login" >>= getJSON 200 >>= asString
+  bindResponse (generateVerificationCode' owner email) $ \resp -> do
+    resp.status `shouldMatchInt` 429
+  mostRecentCode <- retryT $ do
+    resp <- generateVerificationCode' owner email
+    resp.status `shouldMatchInt` 200
+    getVerificationCode owner "login" >>= getJSON 200 >>= asString
+
+  bindResponse (loginWith2ndFactor owner email defPassword fstCode) \resp -> do
+    resp.status `shouldMatchInt` 403
+    resp.json %. "label" `shouldMatch` "code-authentication-failed"
+
+  bindResponse (loginWith2ndFactor owner email defPassword mostRecentCode) \resp -> do
+    resp.status `shouldMatchInt` 200
+
+testLoginVerify6DigitLimitRetries :: HasCallStack => App ()
+testLoginVerify6DigitLimitRetries = do
+  (owner, team, []) <- createTeam OwnDomain 0
+  email <- owner %. "email"
+  setTeamFeatureLockStatus owner team "sndFactorPasswordChallenge" "unlocked"
+  setTeamFeatureStatus owner team "sndFactorPasswordChallenge" "enabled"
+  generateVerificationCode owner email
+  correctCode <- getVerificationCode owner "login" >>= getJSON 200 >>= asString
+  let wrongCode = "123456"
```

Comment:
i guess every one in a million times, this test will probably fail, but that's ok?  :)